### DISCUSSION TITLE
Dates

### DISCUSSION THREAD

**Discussion by jrbarnes9** [ONSdigital/design-system#22] [Feedback]
Use this place to discuss the [dates pattern](https://ons-design-system.netlify.com/patterns/dates) in the [ONS Design System](https://ons-design-system.netlify.com)

**Comment by bameyrick** [comment]
It would be good to be able to set attributes on each input element to add things like autocomplete parameters.

**Reply by jrbarnes9** [reply]
#194 resolved

**Comment by bameyrick** [comment]
The date input example should have a mode to render it as a question, where the legend is a H1 and the description is 1rem with the margin moved to below the description rather than between the legend and description. The individual field labels should also be bold.

**Reply by jrbarnes9** [reply]
#206 resolved

**Comment by bameyrick** [comment]
The date input component wraps itself in a fieldset even if it's only configured to ask for one field (e.g. just the year field)

**Reply by jrbarnes9** [reply]
#517 resolved

**Comment by bameyrick** [comment]
When the date component is configured to only show one field, the fields are incorrectly wrapped in a `field-group`

![Screenshot 2019-07-16 at 15 14 32](https://user-images.githubusercontent.com/8447446/61301737-6fb09d00-a7dc-11e9-9035-ce6996bc4105.png)

**Reply by jrbarnes9** [reply]
#550 resolved

**Comment by jrbarnes9** [comment]
Users have fed back frustrations around the ability to add more than two or four characters in the date inputs.  This is especially frustrating for users who expect auto-tabbing, because they end up entering everything into the first input (which is too small to easily edit/delete)

**Reply by jrbarnes9** [reply]
#672 resolved

**Comment by jrbarnes9** [comment]
We could do with independent field error styling for the date component so error messages can be specific to the invalid field

**Reply by jrbarnes9** [reply]
#1573 raised

**Comment by jrbarnes9** [comment]
Space missing between date fieldset and error panel

<img width="515" alt="Screenshot 2022-01-28 at 17 27 32" src="https://user-images.githubusercontent.com/43346934/151593869-89705c60-1e41-4cd8-8759-05710713ea80.png">

**Reply by jrbarnes9** [reply]
Raised #1896

**Comment by Al-Davies** [comment]
#UR Findings
tested as part of health monitoring survey - usability testing 6-12 December 2022 (testing involved multiple instances where participants had to enter dates for samples and symptom dates). No guidance was given within the survey on how to enter the dates

* All 5 participants entered dates in full i.e. 12.12.2022
* Where error messages were displayed (i.e. warning to enter dates after a specific period) participants understood warning messages and corrected dates accordingly
* When questioned about why they had entered 2022 rather than 22, participants said that the box sizes to enter the year was larger than those for the date and month, therefore leading them to assume they needed to enter a 4 digit year.

FULL DATE

- often users expect to autotab from date to month when completing date entries
- "ah, so this doesn’t move over when I've done the date" "I prefer it if it just moves to the next box"
- "some sites automatically jump to the next field, some don't, but automatic is my preference"

MONTH/YEAR ONLY

- this pattern has been successfully used for Census EQ - where users are asked for date of entry to the UK - where they are only asked for month and year. Many participants remember the year, but may be less sure about the month. Asking for date would likely lead to high inaccuracy.

**Comment by mr-nick17** [comment]
This component contains invalid html and is therefore breaking web standards. The `min` and `max` [attributes are not valid](https://html.spec.whatwg.org/multipage/input.html#the-min-and-max-attributes) on input type of text and considering that browser validation is recommended to be [turned off](https://service-manual.ons.gov.uk/design-system/patterns/correct-errors#turn-off-html5-validation) there's little value in having them. [Self closing tags](https://developer.mozilla.org/en-US/docs/Glossary/Void_element#self-closing_tags) are also invalid `/>` and should be removed.

**Reply by daniellecorke** [reply]
Hi Nick, thank you for your feedback. I have raised #3098 to look into this.

**Reply by daniellecorke** [reply]
Update: Thanks for highlighting this issue for us Nick. 

We have completed #3160 to remove `min` and `max` HTML attributes for inputs with type set to "text". This will be available in our next release. We didn't find invalid html for the self-closing tags and we will now close #3098. 

Please let us know if you have any further feedback.

**Reply by mr-nick17** [reply]
Thanks for the update. I would encourage as part of your investigation, and developer good practise, that the `html` output is run through a validator e.g. https://validator.w3.org/nu/#textarea. Doing so would have highlighted the `min` `max` on the incorrect input type much earlier in the development cycle. With regards to self-closing tags, although technically the html is valid, it is not good practise, by checking the 'date' output (and ignoring the upcoming fix for min/max) this info is supplied with regards to void elements

_Info: Trailing slash on void elements [has no effect](https://github.com/validator/validator/wiki/Markup-%C2%BB-Void-elements#trailing-slashes-in-void-element-start-tags-do-not-mark-the-start-tags-as-self-closing) and [interacts badly with unquoted attribute values](https://github.com/validator/validator/wiki/Markup-%C2%BB-Void-elements#trailing-slashes-directly-preceded-by-unquoted-attribute-values)._

<img width="1523" alt="info on trialling slashes" src="https://github.com/ONSdigital/design-system/assets/19624419/9a321e3f-86ee-4b34-a40c-d3da58d9838d">

**Reply by daniellecorke** [reply]
Thanks for this input Nick, I'll pass it on to the team.
